core-web/libs/portlets/dot-ema/src/lib/edit-ema-editor/components/edit-ema-palette/store/edit-ema-palette.store.ts

```diff
@@ -0,0 +1,201 @@
+    }));
+
+    readonly setStatus = this.updater((state, status: EditEmaPaletteStoreStatus) => ({
+        ...state,
+        status
+    }));
+
+    readonly resetContentlets = this.updater((state) => ({
+        ...state,
+        contentlets: {
+            items: [],
+            filter: { query: '', contentTypeVarName: '' },
+            totalRecords: 0,
+            itemsPerPage: PALETTE_PAGINATOR_ITEMS_PER_PAGE
+        }
+    }));
+
+    /**
+     * Loads content types based on the provided filter and allowed content.
+     * @param data$ An observable that emits an object containing the filter and allowed content.
+     * @returns An observable that emits the loaded content types.
+     */
+    readonly loadContentTypes = this.effect(
+        (data$: Observable<{ filter: string; allowedContent: string[] }>) => {
+            this.resetContentlets();
+
+            return data$.pipe(
+                switchMap(({ allowedContent, filter }) => {
+                    return this.getFilteredContentTypes(filter, allowedContent).pipe(
+                        tapResponse(
+                            (contentTypes) => {
+                                this.patchState({
+                                    contenttypes: { items: contentTypes, filter },
+                                    status: EditEmaPaletteStoreStatus.LOADED
+                                });
+                            },
+                            (err) => {
+                                this.setStatus(EditEmaPaletteStoreStatus.ERROR);
+                                // eslint-disable-next-line no-console
+                                console.log(err);
+                            }
+                        )
+                    );
+                })
+            );
+        }
+    );
+
+    /**
+     * Loads contentlets based on the provided filter, content type name, language ID, and page number.
+     * @param data$ An observable that emits an object containing the filter, content type name, language ID, and page number.
+     * @returns An observable that emits the loaded contentlets.
+     */
+    readonly loadContentlets = this.effect(
+        (
+            data$: Observable<{
+                filter: string;
+                contenttypeName: string;
+                languageId: string;
+                page?: number;
+            }>
+        ) => {
+            return data$.pipe(
+                tap(() => this.setStatus(EditEmaPaletteStoreStatus.LOADING)),
+                switchMap(({ filter, contenttypeName, languageId, page }) => {
+                    return this.dotESContentService
+                        .get({
+                            itemsPerPage: PALETTE_PAGINATOR_ITEMS_PER_PAGE,
+                            lang: languageId || '1',
+                            filter: filter || '',
+                            offset:
+                                ((page || 0) * PALETTE_PAGINATOR_ITEMS_PER_PAGE).toString() || '0',
```

Comment:
This is really confusing. Can you set page to 0 if it is undefined/null?

Reply:
Yeah, confusing... not sure what's going on here, maybe we can move this to a private calculation method and just try to keep the code as clear as possible.